lib/pos_network.go

```diff
@@ -0,0 +1,569 @@
+	if err != nil {
+		return false
+	}
+
+	otherEncodedBytes, err := other.ToBytes()
+	if err != nil {
+		return false
+	}
+
+	return bytes.Equal(sigEncodedBytes, otherEncodedBytes)
+}
+
+func (sig *AggregatedBLSSignature) ToBytes() ([]byte, error) {
+	retBytes := []byte{}
+
+	// SignersList
+	if sig.SignersList == nil {
+		return nil, errors.New("AggregatedBLSSignature.ToBytes: SignersList must not be nil")
+	}
+	retBytes = append(retBytes, EncodeBitset(sig.SignersList)...)
+
+	// Signature
+	if sig.Signature == nil {
+		return nil, errors.New("AggregatedBLSSignature.ToBytes: Signature must not be nil")
+	}
+	retBytes = append(retBytes, EncodeBLSSignature(sig.Signature)...)
+
+	return retBytes, nil
+}
+
+func (sig *AggregatedBLSSignature) FromBytes(rr io.Reader) error {
+	var err error
+
+	sig.SignersList, err = DecodeBitset(rr)
+	if err != nil {
+		return errors.Wrapf(err, "AggregatedBLSSignature.FromBytes: Error decoding SignersList")
+	}
+
+	sig.Signature, err = DecodeBLSSignature(rr)
+	if err != nil {
+		return errors.Wrapf(err, "AggregatedBLSSignature.FromBytes: Error decoding Signature")
+	}
+
+	return nil
+}
+
+// TimeoutAggregateQuorumCertificate is an aggregation of timeout messages from 2/3rds
+// of all validators, weighted by stake, which indicates that these validators want to
+// time out a particular view.
+//
+// When validators want to time out a view, they send their high QCs to the block proposer
+// who builds an aggregate QC extending the chain from the highest QC that it received
+// from all validators who timed out. To prove that it has selected the highest QC, the
+// proposer also includes a list of the high QC views that each validator has sent.
+type TimeoutAggregateQuorumCertificate struct {
+
+	// The view that the block proposers has produced a timeout QC for.
+	TimedOutView uint64
+
+	// This is the highest QC that the block proposer received from any validator who
+	// has timed out for the current view.
+	ValidatorsHighQC *QuorumCertificate
+
+	// Here we include a list of the HighQC.View values we got from each of the
+	// validators in the ValidatorsTimeoutHighQCViews field. In addition, for each
+	// unique HighQC.View value we received, we combine all the partial signatures
+	// for that HighQC.View into a single BLSMultiSignature.
+	//
+	//
+	// The aggregated signature is made up of partial signatures for all present
+	// validators, each of whom signed a payload with the pair
+	// (current view, the validator's local HighQC.View).
+	//
+	// The ordering of high QC views and validators in the aggregate signature will
+	// match the ordering of active validators in descending order of stake for the
+	// timed out view's epoch. I.e. index 0 will correspond to the highest-staked active
+	// validator in the epoch, index 1 will correspond to the second-highest-staked active
+	// validator, ...
+	ValidatorsTimeoutHighQCViews         []uint64
+	ValidatorsTimeoutAggregatedSignature *AggregatedBLSSignature
```

Comment:
This goes with my other comment on aggregating signatures. I think I might be missing something, but if my previous comment's assumptions aren't wrong, then the fields hree seem like they'd be more like the following:
```
ValidatorsTimeoutHighQCViews         []uint64

ValidatorsTimeoutHighQCPublicKeys    []BitSet // Same list size as the previous field. Each bitset in this field represents the pubkeys corresponding to a particular HighQCView.

ValidatorsTimeoutAggregatedSignature *Signature // I think this would be a vanilla "Signature" from signature.go that you would pass to VerifyAggregateSignatureMultiplePayloads
```

So then you would do something like this to verify a timeout qc:
```
// Assume tqc is a TimeoutAggregateQuorumCertificate
publicKeys = [][]*PublicKey{}
utxoView := NewUtxoView() // Have to get a view to convert the bitsets properly
for bs := range tqc.TimeoutAggregateQuorumCertificate {
    publicKeys = append(publicKeys, utxoView.BitSetToPubKeys(bs))
}

payloads = tqc.ValidatorsTimeoutHighQCViews // Maybe these need to be hashed or whatever but you get the idea

verified, err := VerifyAggregateSignatureMultiplePayloads(publicKeys, tqc.ValidatorsTimeoutAggregatedSignature, payloads)
```

Again I feel like I'm missing something. Just typing out what I was thinking so I can get up to speed.

Reply:
Ok so adding to my previous comment, let me confirm my understanding here.

Suppose we have the following situation (notice I reordered some pks to make things interesting):
```
HighQC1 is signed by {pk4, pk5, pk3}
HighQC2 is signed by {pk1, pk2}
HighQC3 is signed by {pk6, pk7, pk8}
```

I think then the fields would be as follows:
```
ValidatorsTimeoutHighQCViews = [2, 2, 1, 1, 1, 3, 3, 3]
ValidatorsTimeoutAggregatedSignature = AggregatedBLSSignature{
    // The bitmap would be set so that it's converted as follows...
    bitmap = [pk1, pk2, pk3, pk4, pk5, pk6, pk7, pk8]
    // Then the signature aggregates (pk1, pk1HighQCView), (pk2, pk2HighQCView), ...
}
```

With the above, the verification would look as follows:
```
VerifyAggregateSignatureSinglePayload([pk1, ..., pk8], signature *Signature, [2, 2, 1, 1, 1, 3, 3, 3])
```

Is that roughly correct?

Reply:
Yup, that's correct!